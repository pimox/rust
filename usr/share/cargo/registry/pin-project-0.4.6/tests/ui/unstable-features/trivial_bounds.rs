// NB: If you change this test, change 'trivial_bounds-feature-gate.rs' at the same time.

// trivial_bounds
// Tracking issue: https://github.com/rust-lang/rust/issues/48214
#![feature(trivial_bounds)]
#![deny(trivial_bounds)]

use std::marker::{PhantomData, PhantomPinned};

fn inner() {
    struct Inner(PhantomPinned);

    struct Foo(Inner);

    impl Unpin for Foo where Inner: Unpin {} //~ ERROR std::marker::Unpin does not depend on any type or lifetime parameters

    struct Wrapper<T>(T);

    impl<T> Unpin for Wrapper<T> where T: Unpin {}

    struct Bar(Inner);

    impl Unpin for Bar where Wrapper<Inner>: Unpin {} //~ ERROR std::marker::Unpin does not depend on any type or lifetime parameters

    struct WrapperWithLifetime<'a, T>(PhantomData<&'a ()>, T);

    impl<T> Unpin for WrapperWithLifetime<'_, T> where T: Unpin {}

    struct Baz(Inner);

    impl<'a> Unpin for Baz where WrapperWithLifetime<'a, Inner>: Unpin {} // Ok
}

fn main() {}
