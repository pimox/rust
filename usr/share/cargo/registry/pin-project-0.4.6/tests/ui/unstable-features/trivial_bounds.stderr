error: Trait bound inner::Inner: std::marker::Unpin does not depend on any type or lifetime parameters
  --> $DIR/trivial_bounds.rs:15:37
   |
15 |     impl Unpin for Foo where Inner: Unpin {} //~ ERROR std::marker::Unpin does not depend on any type or lifetime parameters
   |                                     ^^^^^
   |
note: lint level defined here
  --> $DIR/trivial_bounds.rs:6:9
   |
6  | #![deny(trivial_bounds)]
   |         ^^^^^^^^^^^^^^

error: Trait bound inner::Wrapper<inner::Inner>: std::marker::Unpin does not depend on any type or lifetime parameters
  --> $DIR/trivial_bounds.rs:23:46
   |
23 |     impl Unpin for Bar where Wrapper<Inner>: Unpin {} //~ ERROR std::marker::Unpin does not depend on any type or lifetime parameters
   |                                              ^^^^^
